File: 1036-rotting-oranges/rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        visited:Set[Tuple[int]] = set()

        queue:Deque[Tuple[int]] = deque()

        def addCell(i, j):
            if i < m and i >= 0 and j < n and j >= 0 and grid[i][j] != 0 and not (i,j) in visited:
                visited.add((i,j))
                grid[i][j] = 2
                queue.append((i,j))
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    addCell(i,j)
        
        time:int = -1
        while queue:
            time += 1
            length = len(queue)
            for _ in range(length):
                (x, y) = queue.popleft()
                addCell(x+1, y)
                addCell(x, y+1)
                addCell(x, y-1)
                addCell(x-1, y)
        time = max(0, time)

        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    return -1

        return time

File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:

    bool is_valid(vector<vector<int>>& grid, int x, int y){
        int n = grid.size();
        int m = grid[0].size();
        if(x < 0 || x >= n || y < 0 || y >= m) return false;
        if(grid[x][y] == 0 || grid[x][y] == 2) return false;
        return true;
    }

    int orangesRotting(vector<vector<int>>& grid) {
        queue<pair<int, int>> q;
        int n = grid.size();
        int m = grid[0].size();
        for(int i = 0; i < n;i++){
            for(int j = 0; j < m; j++) {
                if(grid[i][j] == 2) q.push({i, j});
            }
        }
        int time = -1;

        while(!q.empty()) {
            time += 1;
            int q_size = q.size();

            while(q_size--){
                pair<int, int> coords = q.front();q.pop();
                vector<pair<int, int>> directions({{0,1}, {0, -1}, {1, 0}, {-1, 0}});

                for(const auto& dir : directions){
                    int new_x = coords.first + dir.first;
                    int new_y = coords.second + dir.second;
                    if(is_valid(grid, new_x, new_y)){
                        grid[new_x][new_y] = 2;
                        q.push({new_x, new_y});
                    }
                }
            }
        }
        bool all_done = true;

        for(int i = 0; i < n;i++){
            for(int j = 0; j < m; j++) {
                if(grid[i][j] == 1){
                    all_done = false;
                    break;
                }
            }
            if(!all_done) break;
        }
        return all_done?max(time, 0):-1;
    }
};